File: src/service/safety_manager.py
"""Safety manager.

Handles emergency stop, gas alerts, and safe mode transitions.
Priority: EMERGENCY_STOP > SAFE_MODE > MISSION
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum

from domain_types.events import BaseEvent, EventType
from domain_types.terrain import LocomotionMode


class SafetyLevel(IntEnum):
    """Safety levels in priority order (higher = more urgent)."""

    NORMAL = 0
    CAUTION = 1
    SAFE_MODE = 2
    EMERGENCY_STOP = 3


@dataclass
class SafetyState:
    """Current safety system state."""

    level: SafetyLevel = SafetyLevel.NORMAL
    gas_alert_active: bool = False
    emergency_stopped: bool = False
    reason: str = ""
    recommended_locomotion: LocomotionMode = LocomotionMode.WHEEL


class SafetyManager:
    """Manages safety state and produces safety decisions.

    Priority: EMERGENCY_STOP > GAS_ALERT > SAFE_MODE > NORMAL
    """

    def __init__(self) -> None:
        self.state = SafetyState()
        self._event_log: list[BaseEvent] = []

    def handle_event(self, event: BaseEvent) -> SafetyState:
        """Process a safety-relevant event and update state.

        Returns the updated SafetyState.
        """
        self._event_log.append(event)

        if event.event_type == EventType.EMERGENCY_STOP:
            self.state = SafetyState(
                level=SafetyLevel.EMERGENCY_STOP,
                emergency_stopped=True,
                reason=f"Emergency stop: {event.event_id}",
                recommended_locomotion=LocomotionMode.STOP_AND_REPLAN,
            )
        elif event.event_type == EventType.GAS_HAZARD:
            if not self.state.emergency_stopped:
                self.state = SafetyState(
                    level=SafetyLevel.SAFE_MODE,
                    gas_alert_active=True,
                    reason=f"Gas alert: {event.event_id}",
                    recommended_locomotion=LocomotionMode.STOP_AND_REPLAN,
                )
        elif event.event_type == EventType.SLAM_FAILURE:
            if self.state.level < SafetyLevel.SAFE_MODE:
                self.state = SafetyState(
                    level=SafetyLevel.CAUTION,
                    reason=f"SLAM failure: {event.event_id}",
                    recommended_locomotion=LocomotionMode.SLOW_SAFE,
                )
        elif event.event_type == EventType.WEBRTC_DISCONNECTED:
            if self.state.level < SafetyLevel.SAFE_MODE:
                self.state = SafetyState(
                    level=SafetyLevel.CAUTION,
                    reason=f"Network disconnected: {event.event_id}",
                    recommended_locomotion=LocomotionMode.SLOW_SAFE,
                )

        return self.state

    def reset(self) -> None:
        """Reset to normal state (only if not emergency stopped)."""
        if not self.state.emergency_stopped:
            self.state = SafetyState()

    def is_safe_to_operate(self) -> bool:
        """Check if normal operation is allowed."""
        return self.state.level <= SafetyLevel.CAUTION

    def get_event_log(self) -> list[BaseEvent]:
        """Return all processed safety events."""
        return list(self._event_log)

File: go_core/internal/service/safety_manager.go
package service

import (
	"fmt"
	"sync"
	"go_core/internal/domain"
)

type SafetyLevel int

const (
	SafetyLevelNormal        SafetyLevel = 0
	SafetyLevelCaution       SafetyLevel = 1
	SafetyLevelSafeMode      SafetyLevel = 2
	SafetyLevelEmergencyStop SafetyLevel = 3
)

type SafetyState struct {
	Level                 SafetyLevel           `json:"level"`
	GasAlertActive        bool                  `json:"gas_alert_active"`
	EmergencyStopped      bool                  `json:"emergency_stopped"`
	Reason                string                `json:"reason"`
	RecommendedLocomotion domain.LocomotionMode `json:"recommended_locomotion"`
}

type SafetyManager struct {
	mu       sync.RWMutex
	state    SafetyState
	eventLog []domain.BaseEvent
}

func NewSafetyManager() *SafetyManager {
	return &SafetyManager{
		state: SafetyState{
			Level:                 SafetyLevelNormal,
			GasAlertActive:        false,
			EmergencyStopped:      false,
			Reason:                "",
			RecommendedLocomotion: domain.LocomotionModeWheel,
		},
	}
}

func (sm *SafetyManager) HandleEvent(event domain.BaseEvent) SafetyState {
	sm.mu.Lock()
	defer sm.mu.Unlock()

	sm.eventLog = append(sm.eventLog, event)

	if event.EventType == domain.EventTypeEmergencyStop {
		sm.state = SafetyState{
			Level:                 SafetyLevelEmergencyStop,
			EmergencyStopped:      true,
			Reason:                fmt.Sprintf("Emergency stop: %s", event.EventID),
			RecommendedLocomotion: domain.LocomotionModeStopAndReplan,
		}
	} else if event.EventType == domain.EventTypeGasHazard {
		if !sm.state.EmergencyStopped {
			sm.state = SafetyState{
				Level:                 SafetyLevelSafeMode,
				GasAlertActive:        true,
				Reason:                fmt.Sprintf("Gas alert: %s", event.EventID),
				RecommendedLocomotion: domain.LocomotionModeStopAndReplan,
			}
		}
	} else if event.EventType == domain.EventTypeSlamFailure {
		if sm.state.Level < SafetyLevelSafeMode {
			sm.state = SafetyState{
				Level:                 SafetyLevelCaution,
				Reason:                fmt.Sprintf("SLAM failure: %s", event.EventID),
				RecommendedLocomotion: domain.LocomotionModeSlowSafe,
			}
		}
	} else if event.EventType == domain.EventTypeWebRTCDisconnected {
		if sm.state.Level < SafetyLevelSafeMode {
			sm.state = SafetyState{
				Level:                 SafetyLevelCaution,
				Reason:                fmt.Sprintf("Network disconnected: %s", event.EventID),
				RecommendedLocomotion: domain.LocomotionModeSlowSafe,
			}
		}
	}

	return sm.state
}

func (sm *SafetyManager) Reset() {
	sm.mu.Lock()
	defer sm.mu.Unlock()

	if !sm.state.EmergencyStopped {
		sm.state = SafetyState{
			Level:                 SafetyLevelNormal,
			GasAlertActive:        false,
			EmergencyStopped:      false,
			Reason:                "",
			RecommendedLocomotion: domain.LocomotionModeWheel,
		}
	}
}

func (sm *SafetyManager) IsSafeToOperate() bool {
	sm.mu.RLock()
	defer sm.mu.RUnlock()

	return sm.state.Level <= SafetyLevelCaution
}

func (sm *SafetyManager) GetEventLog() []domain.BaseEvent {
	sm.mu.RLock()
	defer sm.mu.RUnlock()

	logCopy := make([]domain.BaseEvent, len(sm.eventLog))
	copy(logCopy, sm.eventLog)
	return logCopy
}
